tests/ga/test_remoteaccess_handler.py

```diff
@@ -0,0 +1,358 @@
+
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="]aPPEv}uNg1FPnl?")
+    def test_add_user_bad_creation_data(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "]aPPEv}uNg1FPnl?"
+        tstuser = ""
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertEqual(0, os_util.all_users.keys().__len__())
+
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="")
+    def test_add_user_bad_password_data(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = ""
+        tstuser = "foobar"
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertEqual(0, os_util.all_users.keys().__len__())
+
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="]aPPEv}uNg1FPnl?")
+    def test_add_user_already_existing(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "]aPPEv}uNg1FPnl?"
+        tstuser = "foobar"
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertTrue(tstuser in os_util.all_users, "{0} missing from users".format(tstuser))
+        self.assertEqual(1, os_util.all_users.keys().__len__())
+        actual_user = os_util.all_users[tstuser]
+        self.assertEqual(actual_user[7], (expiration_date + timedelta(days=1)).strftime("%Y-%m-%d"))
+        # add the new duplicate user, ensure it's not created and does not overwrite the existing user.
+        # this does not test the user add function as that's mocked, it tests processing skips the remaining
+        # calls after the initial failure
+        new_user_expiration = datetime.utcnow() + timedelta(days=5)
+        rah.add_user(tstuser, pwd, new_user_expiration)
+        self.assertTrue(tstuser in os_util.all_users, "{0} missing from users after dup user attempted".format(tstuser))
+        self.assertEqual(1, os_util.all_users.keys().__len__())
+        actual_user = os_util.all_users[tstuser]
+        self.assertEqual(actual_user[7], (expiration_date + timedelta(days=1)).strftime("%Y-%m-%d"))
```

Comment:
Same comment about using the architected interface to check for correct behavior.

Reply:
changed

---
In reply to: [194536077](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194536077) [](ancestors = 194536077)